File: go/stack/150_eval_rpn.go
package stack

import "strconv"

// evalRPN 逆波兰表达式求值
func evalRPN(tokens []string) int {
	// 栈，
	// 碰到 有效的算符为 '+'、'-'、'*' 和 '/' 。 将其前两个元素出栈进行计算
	stack := make([]int, 0)

	for _, v := range tokens {

		num, err := strconv.Atoi(v)
		if err == nil {
			stack = append(stack, num)
		} else {
			// 转化出错说明是有效运算符
			left := stack[len(stack)-2]
			right := stack[len(stack)-1]
			stack = stack[:len(stack)-2]

			var res int
			switch v {
			case "+":
				res = left + right
			case "-":
				res = left - right
			case "*":
				res = left * right
			case "/":
				res = left / right
			}
			stack = append(stack, res)
		}
	}

	return stack[0]
}

File: rust/src/stack/_150_eval_rpn.rs
/// 逆波兰表达式求值
pub fn eval_rpn(tokens: Vec<String>) -> i32 {
    // 栈模拟

    let mut stack = vec![];

    for v in tokens {
        match v.parse::<i64>() {
            Ok(num) => stack.push(num),

            Err(_) => {
                let right = stack.pop().unwrap();
                let left = stack.pop().unwrap();

                let mut res = 0_i64;

                match v.as_str() {
                    "+" => res = left + right,
                    "-" => res = left - right,
                    "*" => res = left * right,
                    "/" => res = left / right,
                    _ => (),
                }
                stack.push(res)
            }
        }
    }

    stack[0] as i32
}
